[patching] Add conadata format and new patch fields (#11713)
conan-io/conan-center-index: docs/conandata_yml_format.md

> # conandata.yml

Reviewer: Documenting this file is absolutely needed, but I would organize it with fewer and longer sections. This is just a suggestion, feel free to ignore it (close/resolve it).

---

# conandata.yml

<the intro 👍 >

In the context of CCI, this file is mandatory and consist in two main sections that we will explain in the next sections with more detail:

 * `sources`: a paragraph to explain the purpose 
 * `patches`: a paragraph to explain the purpose

## sources

(The same paragraph as above) This is the entry that contains all the items that are downloaded from internet and used by a recipe. This section contains one entry per each version and, each version should declare its own sources.

This is a basic example of a regular library, it should satisfy most of the use cases:

```yml
sources:
   "1.2.11":
       url: "..."
       sha256: "..."
   "1.2.12":
       url: "..."
       sha256: "..."
```

Every entry for a version consists in a dictionary with the `url` and the hashing algorithm of the artifact. `sha256` is required, but others like `sha1` or `md5` can be used as well.

**Mirrors.-** Sometimes it is useful to declare mirrors, use a list in the `url` field. Conan will try to download the artifacts from any of those mirrors.

```yml
sources:
   "1.2.11":
       url:
        - "url1"
        - "url2"
       sha256: "..."
```

## patches

Sometimes it is needed to apply some modifications to downloaded sources... blabalbla...

This section follows the same pattern as the `sources` above: one entry per version with a list of patches to apply. 

```yaml
patches:
   "1.2.11":
     - patch_file: ...
       base_path: ...
       url: ...
       sha256: ...
       patch_type:  ...
       patch_source: ...
       patch_description: ..
```

These fields have the following purpose (using _bold_ for those that are required):
 * `patch_file`: ...
 * `base_path`: ...
 * ...

## Some special cases

Sometimes there are some libraries that need to download several artifacts from internet, or the sources are different for each configuration,... here you can find how to declare these use cases:

### Several artifacts per version

Example: source code + license file

```yml
sources:
  8.0.0:
    - url: https://github.com/approvals/ApprovalTests.cpp/releases/download/v.8.0.0/ApprovalTests.v.8.0.0.hpp
      sha256: e16a97081f8582be951d95a9d53dc611f1f5a84e117a477029890d0b34ae99d6
    - url: "https://raw.githubusercontent.com/approvals/ApprovalTests.cpp/v.8.0.0/LICENSE"
      sha256: c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4
```

### Different sources per configuration

```yml
sources:
  "0066":
    "Macos":
      "apple-clang":
        "x86":
          - url: "https://naif.jpl.nasa.gov/pub/naif/misc/toolkit_N0066/C/MacIntel_OSX_AppleC_32bit/packages/cspice.tar.Z"
            sha256: "9a4b5f674ea76821c43aa9140829da4091de646ef3ce40fd5be1d09d7c37b6b3"
        "x86_64":
          - url: "https://naif.jpl.nasa.gov/pub/naif/misc/toolkit_N0066/C/MacIntel_OSX_AppleC_64bit/packages/cspice.tar.Z"
            sha256: "f5d48c4b0d558c5d71e8bf6fcdf135b0943210c1ff91f8191dfc447419a6b12e"
```

`conandata.yml` is a built-in Conan feature (available since 1.22.0) without a fixed structure, but conan-center-index uses it for its own purposes.

## sources

`sources` is a top level dictionary, containing entries of sources and checksums for each of the supported versions.

### url

`url` contains a string specifying [URI](https://tools.ietf.org/html/rfc3986) where to download released sources.
Usually, `url` has a [https](https://tools.ietf.org/html/rfc2660) scheme, but other schemes, such as [ftp](https://tools.ietf.org/html/rfc959) are accepted as well.

### sha256

[sha256](https://tools.ietf.org/html/rfc6234) is a preferred method to specify hash sum for the released sources. It allows to check the integrity of sources downloaded.
You may use an [online service](https://hash.online-convert.com/sha256-generator) to compute `sha256` sum for the given `url`.
Also, you may use [sha256sum](https://linux.die.net/man/1/sha256sum) command ([windows](http://www.labtestproject.com/files/win/sha256sum/sha256sum.exe)).

### sha1

[sha1](https://tools.ietf.org/html/rfc3174) is an alternate method to specify hash sum. It's usage is discouraged and `sha256` is preferred.

### md5

> [md5](https://tools.ietf.org/html/rfc1321) is an alternate method to specify hash sum. It's usage is discouraged and `sha256` is preferred.

Reviewer 2: Yes!

### sources - Different use cases

There are several ways to specify sources.

#### Single source

This is the most straightforward format:

```
sources:
  "1.18.0":
    sha256: 820688d1e0387ff55194ae20036cbae0fb3c7d11b7c3f46492369723c01df96f
    url: https://github.com/chriskohlhoff/asio/archive/asio-1-18-0.tar.gz
```

#### Single source with download mirror

Sometimes it's a good idea to specify several download mirrors, as some hosts are known to be unreliable or banned in certain countries:

```
sources:
  "1.2.11":
    url: [
         "https://zlib.net/zlib-1.2.11.tar.gz",
         "https://downloads.sourceforge.net/project/libpng/zlib/1.2.11/zlib-1.2.11.tar.gz",
       ]
    sha256: "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1"
```

Keep in mind all the mirrors have to provide the exactly same source (e.g. no repackaging), thus using the same hash sum.

#### Source code + license

> Certain projects provide license on its own, and released artifacts do not include it. In this case, a license URL can be provided separately:

Reviewer 3: ```suggestion
Certain projects provide license on their own, and released artifacts do not include it. In this case, a license URL can be provided separately:
```

### patch_file

Patch file might be committed to the conan-center-index, near to the conanfile (usually, into the `patches` sub-directory). Such patch files usually have either `.diff` or `.patch` extension.
The recommended way to generate such patches is [git format-patch](https://git-scm.com/docs/git-format-patch). The path to the patch is relative to the directory containing `conandata.yml` and `conanfile.py`.

### base_path

Specifies a sub-directory in project's sources to apply patch. This directory is relative to the [source_folder](https://docs.conan.io/en/latest/reference/conanfile/attributes.html?highlight=source_folder#source-folder). Usually, it would be a `source_subfolder`, but could be a lower-level sub-directory (e.g. if it's a patch for a submodule).

### url

The patch could be taken from the existing resource (e.g. some patches are distributed alongside the source tarball). See `url` section of the `sources`.

### sha256

This is the hash for the patch itself, in the same way this field is used in the `sources` section.

### sha1

This is the hash for the patch itself, in the same way this field is used in the `sources` section.

### md5

> This is the hash for the patch itself, in the same way this field is used in the `sources` section.

Reviewer 4: The idea of us downloading patches from another source is new. Was there so far an actual use case of this?

Anyway, I would prefer to limit this to `sha256` from the beginning

### patch_type

The `patch_type` field specifies the type of the patch. In conan-center-index we currently accept only several kind of patches:

#### official

`patch_type: official` indicates the patch is distributed with source code itself. usually, this happens if release managers failed to include a critical fix to the release, but it's too much burden for them to make a new release just because of that single fix.
[example](https://www.boost.org/users/history/version_1_72_0.html) (notice the `coroutine` patch)

#### vulnerability

`patch_type: vulnerability`: Indicates a patch that addresses the security issue. The patch description
should include the index of CVE or CWE the patch addresses.

> Usually, original library projects do new releases fixing vulnerabilities this kind of issues, but in some cases they are either abandoned or inactive.

Reviewer 3: ```suggestion
Usually, original library projects do new releases fixing vulnerabilities for this kind of issues, but in some cases they are either abandoned or inactive.
```

        "x86":
          - url: "https://naif.jpl.nasa.gov/pub/naif/misc/toolkit_N0066/C/MacIntel_OSX_AppleC_32bit/packages/cspice.tar.Z"
            sha256: "9a4b5f674ea76821c43aa9140829da4091de646ef3ce40fd5be1d09d7c37b6b3"
        "x86_64":
          - url: "https://naif.jpl.nasa.gov/pub/naif/misc/toolkit_N0066/C/MacIntel_OSX_AppleC_64bit/packages/cspice.tar.Z"
            sha256: "f5d48c4b0d558c5d71e8bf6fcdf135b0943210c1ff91f8191dfc447419a6b12e"
```

This approach requires a special code within [build](https://docs.conan.io/en/latest/reference/conanfile/methods.html#build) method to handle.

## patches

Sometimes sources provided by project require patching for various reasons. The `conandata.yml` file is the right place to indicate this information as well.

This section follows the same pattern as the `sources` above: one entry per version with a list of patches to apply.

```yaml
patches:
  "1.2.0":
    - patch_file: "patches/1.2.0-002-link-core-with-find-library.patch"
      base_path: "source_subfolder"
      url: "https://a-url-to-a-pull-request-mail-list-topic-issue-or-question"
      sha256: "qafe4rq54533qa43esdaq53ewqa5"
      patch_type: "portability"

>       patch_source: ...

Author: What is the difference between `url` and `patch_source` fields?

Reviewer 3: `url` is URI for patch file itself (.diff or .patch format)
`patch_source` might be an URL to a JIRA ticker, mail list thread or similar thing
it's explained below I guess

Reviewer 5: Does that mean the example is wrong?

Would a practical example be...

```yaml
url: "https://github.com/conan-io/conan-center-index/pull/11713.diff"
patch_source: "https://github.com/conan-io/conan-center-index/pull/11713"
```

or maybe 

```yaml
url: "https://www.boost.org/patches/1_79_0/0001-json-array-erase-relocate.patch"
patch_source: "https://github.com/boostorg/json/issues/692"
```

It's hard to image another situation where the `url` is different from the `patch_source`

Reviewer 3: yes, example requires a small fix.
in general, `url` and `patch_source` are going to be different in 99% of cases, as `patch_source` points to some discussion or bug report.

Author: I think this is too much to ask contributors for...

Reviewer 5: > I think this is too much to ask contributors for...

I agree 🤔

      patch_description: "Link CoreFoundation and CoreServices with find_library"
```

### Patches fields

#### patch_file

Patch file might be committed to the conan-center-index, near to the conanfile (usually, into the `patches` sub-directory). Such patch files usually have either `.diff` or `.patch` extension.
The recommended way to generate such patches is [git format-patch](https://git-scm.com/docs/git-format-patch). The path to the patch is relative to the directory containing `conandata.yml` and `conanfile.py`.

#### base_path

Specifies a sub-directory in project's sources to apply patch. This directory is relative to the [source_folder](https://docs.conan.io/en/latest/reference/conanfile/attributes.html?highlight=source_folder#source-folder). Usually, it would be a `source_subfolder`, but could be a lower-level sub-directory (e.g. if it's a patch for a submodule).

#### url

> The patch could be taken from the existing resource (e.g. some patches are distributed alongside the source tarball). See `url` section of the `sources`.

Reviewer 5: Trying to capture https://github.com/conan-io/conan-center-index/pull/11713/files#r923287561 since I found it a little unclear

```suggestion
When the patch is taken from an existing resource (e.g. some patches are distributed alongside the source tarball) this should point to the same `<filename>.<extension>` as used in the `patch_file`. [`patch_source`](#patch-source) may point to the same resource or to a more human-friendly view. See `url` section of the `sources` for details.
```

#### patch_type

The `patch_type` field specifies the type of the patch. In conan-center-index we currently accept only several kind of patches:

##### official

`patch_type: official` indicates the patch is distributed with source code itself. usually, this happens if release managers failed to include a critical fix to the release, but it's too much burden for them to make a new release just because of that single fix.
[example](https://www.boost.org/users/history/version_1_72_0.html) (notice the `coroutine` patch)

##### vulnerability

`patch_type: vulnerability`: Indicates a patch that addresses the security issue. The patch description
should include the index of CVE or CWE the patch addresses.
Usually, original library projects do new releases fixing vulnerabilities for this kind of issues, but in some cases they are either abandoned or inactive.

##### backport

`patch_type: backport`: Indicates a patch that backports an existing bug fix from the newer release or master branch (or equivalent, such as main/develop/trunk/etc). The patch source may be a pull request, or bug within the project's issue tracker.

> Backports are accepted only for bugs, but not for feature requests.

Reviewer: ```suggestion
Backports are accepted only for bugs that break normal execution flow, never for feature requests.
```

This is what we called _broken execution_, it is not just a bug people might be relying on its _buggy_ behavior

For the `patch_type: portability` there might be no patch source matching the definition above. Although we encourage contributors to submit all such portability fixes upstream first, it's not always possible (e.g. for projects no longer maintained). In that case, a link to the Conan issue is a valid patch source (if there is no issue, you may [create](https://github.com/conan-io/conan-center-index/issues/new/choose) one).
For the `patch_type: conan`, it doesn't make sense to submit patch upstream, so there will be no patch source.

#### base_path

_Optional_

Specifies a sub-directory in project's sources to apply patch. This directory is relative to the [source_folder](https://docs.conan.io/en/latest/reference/conanfile/attributes.html?highlight=source_folder#source-folder). Usually, it would be a `source_subfolder`, but could be a lower-level sub-directory (e.g. if it's a patch for a submodule).

#### sha256

_Optional_

This is the hash for the patch itself, in the same way this field is used in the `sources` section.

> 
> #### sha1
> 
> _Optional_
> 
> This is the hash for the patch itself, in the same way this field is used in the `sources` section.
> 
> #### md5
> 
> _Optional_
> 
> This is the hash for the patch itself, in the same way this field is used in the `sources` section.

Reviewer 5: ```suggestion
```
